File: skywall/core/signals.py
import weakref


class Signal:

    def __init__(self, name):
        self.name = name
        self.listeners = []

    def connect(self, listener):
        if hasattr(listener, '__self__'):
            listener_ref = weakref.WeakMethod(listener)
        else:
            listener_ref = weakref.ref(listener)
        self.listeners.append(listener_ref)
        return listener

    def disconnect(self, listener):
        self.cleanup_lost_refs()
        self.listeners = [listener_ref for listener_ref in self.listeners if listener_ref() != listener]

    def emit(self, **kwargs):
        self.cleanup_lost_refs()
        for listener_ref in self.listeners:
            listener = listener_ref()
            if listener is not None:
                listener(**kwargs)

    def cleanup_lost_refs(self):
        self.listeners = [listener_ref for listener_ref in self.listeners if listener_ref() is not None]

File: frontend/utils/signals.js
import {pull} from 'lodash'


export class Signal {
  constructor(name) {
    this.name = name
    this.listeners = []
  }
  connect(listener) {
    this.listeners.push(listener)
    return listener
  }
  disconnect(listener) {
    pull(this.listeners, listener)
  }
  emit(arg) {
    for (const listener of this.listeners) {
      listener(arg)
    }
  }
}

export class FlowSignal extends Signal {
  emit(arg) {
    let res = arg
    for (const listener of this.listeners) {
      res = listener(res)
    }
    return res
  }
}
